samza-sql-shell/src/main/java/org/apache/samza/sql/client/impl/SamzaExecutor.java

```diff
@@ -0,0 +1,582 @@
+
+    private static final String SAMZA_SYSTEM_LOG = "log";
+    private static final String SAMZA_SYSTEM_KAFKA = "kafka";
+    private static final String SAMZA_SQL_OUTPUT = "samza.sql.output";
+    private static final String SAMZA_SQL_SYSTEM_KAFKA_ADDRESS = "samza.sql.system.kafka.address";
+    private static final String DEFAULT_SERVER_ADDRESS = "localhost:2181";
+    private static final int RANDOM_ACCESS_QUEUE_CAPACITY = 5000;
+
+    private static RandomAccessQueue<OutgoingMessageEnvelope> m_outputData =
+        new RandomAccessQueue<>(OutgoingMessageEnvelope.class, RANDOM_ACCESS_QUEUE_CAPACITY);
+    private static AtomicInteger m_execIdSeq = new AtomicInteger(0);
+    private Map<Integer, SamzaSqlApplicationRunner> m_executions = new HashMap<>();
+    private String m_lastErrorMsg = "";
+
+    // -- implementation of SqlExecutor ------------------------------------------
+
+    @Override
+    public void start(ExecutionContext context) {
+    }
+
+    @Override
+    public void stop(ExecutionContext context) {
+        Iterator<Integer> iter =  m_executions.keySet().iterator();
+        while(iter.hasNext()) {
+            stopExecution(context, iter.next());
+            iter.remove();
+        }
+        m_outputData.clear();
+    }
+
+    @Override
+    public List<String> listTables(ExecutionContext context) {
+        /**
+         * TODO: currently the Shell can only talk to Kafka system, but we should use a general way
+         *       to connect to different systems.
+         */
+        String address = context.getConfigMap().getOrDefault(SAMZA_SQL_SYSTEM_KAFKA_ADDRESS,
+            DEFAULT_SERVER_ADDRESS);
+        ZkUtils zkUtils = new ZkUtils(new ZkClient(address), new ZkConnection(address), false);
+        List<String> tables = JavaConversions.seqAsJavaList(zkUtils.getAllTopics())
+            .stream()
+            .map(x -> SAMZA_SYSTEM_KAFKA + "." + x)
+            .collect(Collectors.toList());
+        return tables;
+    }
+
+    @Override
+    public SqlSchema getTableSchema(ExecutionContext context, String tableName) {
+        m_lastErrorMsg = "";
+        int execId = m_execIdSeq.incrementAndGet();
+        Map<String, String> staticConfigs = fetchSamzaSqlConfig(execId, context);
+        Config samzaSqlConfig = new MapConfig(staticConfigs);
+        SqlSchema sqlSchema = null;
+        try {
+            SqlIOResolver ioResolver = SamzaSqlApplicationConfig.createIOResolver(samzaSqlConfig);
+            SqlIOConfig sourceInfo = ioResolver.fetchSourceInfo(tableName);
+            RelSchemaProvider schemaProvider =
+                initializePlugin("RelSchemaProvider", sourceInfo.getRelSchemaProviderName(), samzaSqlConfig,
+                    CFG_FMT_REL_SCHEMA_PROVIDER_DOMAIN,
+                    (o, c) -> ((RelSchemaProviderFactory) o).create(sourceInfo.getSystemStream(), c));
+            AvroRelSchemaProvider avroSchemaProvider = (AvroRelSchemaProvider) schemaProvider;
```

Comment:
Can you add a comment here saying that Shell right now works only for systems that has AvroSchemas?

Reply:
Sure. Done.